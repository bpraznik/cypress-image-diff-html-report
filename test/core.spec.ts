// import { jest } from '@jest/globals'
import fs from 'fs-extra'
import mock from 'mock-fs'
import path from 'path'
import request from 'supertest'
import { generate, startServer } from '../dist/index.js'
import http from 'http'
import { AddressInfo } from 'net'

describe('Core', () => {
  beforeEach(() => {
    mock({
      'dist/ui/index.html': mock.load('dist/ui/index.html'),
      'cypress-image-diff-html-report.config.js': mock.load(
        'test/fixtures/cypress-image-diff-html-report.config.js'
      ),
      'cypress-image-diff-html-report.input.json': mock.load(
        'test/fixtures/cypress-image-diff-html-report.input.json'
      ),
      'my-config.js': mock.load('test/fixtures/my-config.js'),
      'test.json': mock.load('test/fixtures/test.json')
    })
  })

  afterEach(() => {
    mock.restore()
  })

  describe('Generate HTML report', () => {
    it('should create HTML report with given inputJsonPath', async () => {
      try {
        await generate({ inputJsonPath: 'test.json' })
      } catch (e) {
        /* empty */
      }

      const expectedPath = path.join(
        process.cwd(),
        'cypress-image-diff-html-report/index.html'
      )
      expect(fs.existsSync(expectedPath)).toBe(true)

      const expectedHtml = await fs.readFile(expectedPath, { encoding: 'utf8' })
      expect(expectedHtml).toMatchSnapshot()
    })

    it('should create HTML report with given outputDir', async () => {
      try {
        await generate({ outputDir: 'my-report/html' })
      } catch {
        /* empty */
      }

      const expectedPath = path.join(process.cwd(), 'my-report/html/index.html')
      expect(fs.existsSync(expectedPath)).toBe(true)

      const expectedHtml = await fs.readFile(expectedPath, { encoding: 'utf8' })
      expect(expectedHtml).toMatchSnapshot()
    })
  })

  describe('Start the local server', () => {
    let server = new http.Server()

    afterEach(() => {
      server.close()
    })

    it('should serve /', async () => {
      try {
        server = await startServer()
      } catch {
        /* empty */
      }
      const response = await request(server).get('/')
      expect(response.statusCode).toBe(200)
    })

    it('should serve /api/reports', async () => {
      try {
        server = await startServer()
      } catch {
        /* empty */
      }
      const response = await request(server).get('/api/reports')
      expect(response.statusCode).toBe(200)
    })

    it('should start local server with given inputJsonPath', async () => {
      try {
        server = await startServer({ inputJsonPath: 'test.json' })
      } catch {
        /* empty */
      }

      const response = await request(server).get('/api/reports')

      expect(response.statusCode).toBe(200)
      expect(response.type).toBe('application/json')
      expect(response.body).toMatchSnapshot()
    })

    it('should start local server with given autoOpen', async () => {
      // try {
      //   server = await startServer({ autoOpen: true })
      // } catch {
      //   /* empty */
      // }
      // expect(mockedOpen).toHaveBeenCalledTimes(1)
    })

    it('should start local server with given serverPort', async () => {
      try {
        server = await startServer({ serverPort: 4001 })
      } catch {
        /* empty */
      }

      expect((server.address() as AddressInfo).port).toBe(4001)
    })
  })
})
