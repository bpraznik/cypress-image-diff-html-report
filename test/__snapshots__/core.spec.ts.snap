// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core Generate HTML report should create HTML report with given reportJsonFilePath 1`] = `"{"report":{"total":6,"totalPassed":3,"totalFailed":3,"totalSuites":3,"suites":[{"name":"image-diff.cy.js","path":"cypress/specs/image-diff.cy.js","tests":[{"status":"fail","name":"image-diff.cy-wholePage","percentage":0.0900966398590363,"failureThreshold":0,"specPath":"cypress/specs/image-diff.cy.js","specFilename":"image-diff.cy.js","baselinePath":"../cypress-visual-screenshots/baseline/image-diff.cy-wholePage.png","diffPath":"../cypress-visual-screenshots/diff/image-diff.cy-wholePage.png","comparisonPath":"../cypress-visual-screenshots/comparison/image-diff.cy-wholePage.png","passed":0,"failed":1},{"status":"pass","name":"image-diff.cy-wholePageThreshold","percentage":0.0900966398590363,"failureThreshold":0.2,"specPath":"cypress/specs/image-diff.cy.js","specFilename":"image-diff.cy.js","baselinePath":"../cypress-visual-screenshots/baseline/image-diff.cy-wholePageThreshold.png","diffPath":"","comparisonPath":"../cypress-visual-screenshots/comparison/image-diff.cy-wholePageThreshold.png","passed":1,"failed":0},{"status":"pass","name":"image-diff.cy-element","percentage":0,"failureThreshold":0,"specPath":"cypress/specs/image-diff.cy.js","specFilename":"image-diff.cy.js","baselinePath":"../cypress-visual-screenshots/baseline/image-diff.cy-element.png","diffPath":"","comparisonPath":"../cypress-visual-screenshots/comparison/image-diff.cy-element.png","passed":1,"failed":0},{"status":"fail","name":"image-diff.cy-hideElement","percentage":0.0900966398590363,"failureThreshold":0,"specPath":"cypress/specs/image-diff.cy.js","specFilename":"image-diff.cy.js","baselinePath":"../cypress-visual-screenshots/baseline/image-diff.cy-hideElement.png","diffPath":"../cypress-visual-screenshots/diff/image-diff.cy-hideElement.png","comparisonPath":"../cypress-visual-screenshots/comparison/image-diff.cy-hideElement.png","passed":0,"failed":1}],"id":"cypress/specs/image-diff.cy.js","passed":2,"failed":2},{"name":"retry.cy.js","path":"cypress/specs/retry.cy.js","tests":[{"status":"pass","name":"retry.cy-retry","percentage":0,"failureThreshold":0,"specPath":"cypress/specs/retry.cy.js","specFilename":"retry.cy.js","baselinePath":"../cypress-visual-screenshots/baseline/retry.cy-retry.png","diffPath":"","comparisonPath":"../cypress-visual-screenshots/comparison/retry.cy-retry.png","passed":1,"failed":0}],"id":"cypress/specs/retry.cy.js","passed":1,"failed":0},{"name":"folder-structure.cy.js","path":"cypress/specs/folder-structure-test/folder-structure.cy.js","tests":[{"status":"fail","name":"folder-structure.cy-wholePage","percentage":0.0900966398590363,"failureThreshold":0,"specPath":"cypress/specs/folder-structure-test/folder-structure.cy.js","specFilename":"folder-structure.cy.js","baselinePath":"../cypress-visual-screenshots/baseline/folder-structure.cy-wholePage.png","diffPath":"../cypress-visual-screenshots/diff/folder-structure.cy-wholePage.png","comparisonPath":"../cypress-visual-screenshots/comparison/folder-structure.cy-wholePage.png","passed":0,"failed":1}],"id":"cypress/specs/folder-structure-test/folder-structure.cy.js","passed":0,"failed":1}],"startedAt":"2023-09-19T13:15:11.806Z","endedAt":"2023-09-19T13:15:31.112Z","duration":15239,"browserName":"chrome","browserVersion":"116.0.5845.187","cypressVersion":"10.8.0"},"mode":"static"}"`;

exports[`Core Start the local server should start local server with given reportJsonFilePath 1`] = `
{
  "browserName": "chrome",
  "browserVersion": "116.0.5845.187",
  "cypressVersion": "10.8.0",
  "duration": 15239,
  "endedAt": "2023-09-19T13:15:31.112Z",
  "startedAt": "2023-09-19T13:15:11.806Z",
  "suites": [
    {
      "failed": 2,
      "id": "cypress/specs/image-diff.cy.js",
      "name": "image-diff.cy.js",
      "passed": 2,
      "path": "cypress/specs/image-diff.cy.js",
      "tests": [
        {
          "baselinePath": "cypress-visual-screenshots/baseline/image-diff.cy-wholePage.png",
          "comparisonPath": "cypress-visual-screenshots/comparison/image-diff.cy-wholePage.png",
          "diffPath": "cypress-visual-screenshots/diff/image-diff.cy-wholePage.png",
          "failed": 1,
          "failureThreshold": 0,
          "name": "image-diff.cy-wholePage",
          "passed": 0,
          "percentage": 0.0900966398590363,
          "specFilename": "image-diff.cy.js",
          "specPath": "cypress/specs/image-diff.cy.js",
          "status": "fail",
        },
        {
          "baselinePath": "cypress-visual-screenshots/baseline/image-diff.cy-wholePageThreshold.png",
          "comparisonPath": "cypress-visual-screenshots/comparison/image-diff.cy-wholePageThreshold.png",
          "diffPath": "",
          "failed": 0,
          "failureThreshold": 0.2,
          "name": "image-diff.cy-wholePageThreshold",
          "passed": 1,
          "percentage": 0.0900966398590363,
          "specFilename": "image-diff.cy.js",
          "specPath": "cypress/specs/image-diff.cy.js",
          "status": "pass",
        },
        {
          "baselinePath": "cypress-visual-screenshots/baseline/image-diff.cy-element.png",
          "comparisonPath": "cypress-visual-screenshots/comparison/image-diff.cy-element.png",
          "diffPath": "",
          "failed": 0,
          "failureThreshold": 0,
          "name": "image-diff.cy-element",
          "passed": 1,
          "percentage": 0,
          "specFilename": "image-diff.cy.js",
          "specPath": "cypress/specs/image-diff.cy.js",
          "status": "pass",
        },
        {
          "baselinePath": "cypress-visual-screenshots/baseline/image-diff.cy-hideElement.png",
          "comparisonPath": "cypress-visual-screenshots/comparison/image-diff.cy-hideElement.png",
          "diffPath": "cypress-visual-screenshots/diff/image-diff.cy-hideElement.png",
          "failed": 1,
          "failureThreshold": 0,
          "name": "image-diff.cy-hideElement",
          "passed": 0,
          "percentage": 0.0900966398590363,
          "specFilename": "image-diff.cy.js",
          "specPath": "cypress/specs/image-diff.cy.js",
          "status": "fail",
        },
      ],
    },
    {
      "failed": 0,
      "id": "cypress/specs/retry.cy.js",
      "name": "retry.cy.js",
      "passed": 1,
      "path": "cypress/specs/retry.cy.js",
      "tests": [
        {
          "baselinePath": "cypress-visual-screenshots/baseline/retry.cy-retry.png",
          "comparisonPath": "cypress-visual-screenshots/comparison/retry.cy-retry.png",
          "diffPath": "",
          "failed": 0,
          "failureThreshold": 0,
          "name": "retry.cy-retry",
          "passed": 1,
          "percentage": 0,
          "specFilename": "retry.cy.js",
          "specPath": "cypress/specs/retry.cy.js",
          "status": "pass",
        },
      ],
    },
    {
      "failed": 1,
      "id": "cypress/specs/folder-structure-test/folder-structure.cy.js",
      "name": "folder-structure.cy.js",
      "passed": 0,
      "path": "cypress/specs/folder-structure-test/folder-structure.cy.js",
      "tests": [
        {
          "baselinePath": "cypress-visual-screenshots/baseline/folder-structure.cy-wholePage.png",
          "comparisonPath": "cypress-visual-screenshots/comparison/folder-structure.cy-wholePage.png",
          "diffPath": "cypress-visual-screenshots/diff/folder-structure.cy-wholePage.png",
          "failed": 1,
          "failureThreshold": 0,
          "name": "folder-structure.cy-wholePage",
          "passed": 0,
          "percentage": 0.0900966398590363,
          "specFilename": "folder-structure.cy.js",
          "specPath": "cypress/specs/folder-structure-test/folder-structure.cy.js",
          "status": "fail",
        },
      ],
    },
  ],
  "total": 6,
  "totalFailed": 3,
  "totalPassed": 3,
  "totalSuites": 3,
}
`;
